### PR TITLE
haxe.CallStack.exceptionStack for js

### PR DESCRIPTION
**Pull request by nadako**
See #2854 and #3062. This wraps thrown objects that are not instances of `js.Error` in its special subclass, and saves in catch so we can get the stack trace for `haxe.CallStack.exceptionStack`.

```haxe
class Main {
    static function main() {
        try {
            throw false;
        } catch (e:Dynamic) {
            trace(e);
        }
    }
}
```
now generates
```js
Main.main = function() {
	try {
		throw new js__$Boot_HaxeError(false);
	} catch( e ) {
		if (e instanceof js__$Boot_HaxeError) e = e.val;
		console.log(e);
	}
};
```
but only if we did any wraping, so e.g.
```haxe
class Main {
    static function main() {
        try {
            throw new js.Error();
        } catch (e:Dynamic) {
            trace(e);
        }
    }
}
```
generates simple
```js
Main.main = function() {
	try {
		throw new Error();
	} catch( e ) {
		console.log(e);
	}
};
```

now if we use `haxe.Callstack.exceptionStack`
```haxe
class Main {
    static function main() {
        try {
            throw new js.Error();
        } catch (e:Dynamic) {
            var stack = haxe.CallStack.exceptionStack();
            trace(stack);
        }
    }
}
```
we get the following
```js
Main.main = function() {
	try {
		throw new Error();
	} catch( e ) {
		haxe_CallStack.lastException = e;
		var stack = haxe_CallStack.exceptionStack();
		console.log(stack);
	}
};
```

finally, if we throw non-`Error` object and use `exceptionStack`, we generate everything

```haxe
class Main {
    static function main() {
        try {
            throw false;
        } catch (e:Dynamic) {
            var stack = haxe.CallStack.exceptionStack();
            trace(stack);
        }
    }
}
```

```js
Main.main = function() {
	try {
		throw new js__$Boot_HaxeError(false);
	} catch( e ) {
		haxe_CallStack.lastException = e;
		if (e instanceof js__$Boot_HaxeError) e = e.val;
		var stack = haxe_CallStack.exceptionStack();
		console.log(stack);
	}
};
```

One potential problem with this implementation is that if we access `e.stack` before calling `CallStack.exceptionStack()`, the latter will skip custom `prepareStackTrace` function and return a simple `StackItem` array formed by fallback implementation in `CallStack,makeStack`. This may or may not be a problem, but I don't really have a solution besides generating `StackItem`s at the catch point which will hurt performance if we don't use `exceptionStack`, so I think this is not an option.

### CODE REVIEW

**Comment by nadako** [comment]
travis currently fails because of some unrelated as3 issue

**Comment by Simn** [comment]
Oh, I didn't realize it was actually being serious about that for once. :)

File: filters.ml [hunk 1]
```diff
@@ -1066,6 +1088,7 @@ let run com tctx main =
 			captured_vars com;
 			promote_complex_rhs com;
 			if com.config.pf_add_final_return then add_final_return else (fun e -> e);
+			if com.platform = Js then wrap_js_exceptions com else (fun e -> e);
```
Comment (Simn) [inline filters.ml]: Why does that have to be a filter? Couldn't we just use this when generating a TThrow in genjs?

Comment (nadako) [inline filters.ml]: I wanted to run that before the DCE, so HaxeError isn't generated if we don't throw non-`Error` exceptions.

Actually, I think we should have more platform-specific filters that run before the dce, so we can remove unnecesarry `@:keep`s and make generated code much cleaner. 
What I'm not sure about is how to organize that without making a second gencommon.

**Comment by nadako** [comment]
any feedback?

**Comment by aduros** [comment]
Would this break any common interop between pure JS and Haxe? The only thing I can think of is JS that calls into Haxe and expects non-Error types to be thrown. That should be pretty rare.

Is the HaxeError support code only included in the output if needed?

**Comment by nadako** [comment]
Yeah, it should, that's one of the reason it's run as a filter before the DCE.

**Comment by ncannasse** [comment]
Ah, I didn't remember about this one. Should we merge it for 3.2 ?

**Comment by nadako** [comment]
I don't know. I personally have no urgent need for that, but IIRC things like @ciscoheat's buddy library actually wants `CallStack.exceptionStack` to work.

**Comment by ciscoheat** [comment]
@nadako You have good memory. :) Yes I'm using it to make a custom stack view when a test fails: https://github.com/ciscoheat/buddy/blob/master/src/buddy/reporting/TraceReporter.hx#L69-L73

**Comment by nadako** [comment]
We (at the company i work at) came to think that we're actually missing stacktraces from js but do like throwing simple strings :-P So I'm in favor of merging this PR (I'd like to see some comments about potential prolem I mentioned in issue description).

**Comment by Simn** [comment]
Can't be merged as-is.

**Comment by nadako** [comment]
Oh, right, @ncannasse also changed something wrt `haxe.CallStack`. I'll update the PR.

**Comment by nadako** [comment]
@ncannasse could you please check IE10 stuff you worked on on this branch?

File: filters.ml (hunk 1 above)
Comment (Simn) [inline filters.ml]: We have problems with this for python as well, so I'm fine with adding target-specific filters there. I don't think we have to abstract this in any way, so what you did there is fine.

**Comment by frabbit** [comment]
would love to see this feature implemented for 3.2, any objections? @ncannasse @simn

**Comment by ncannasse** [comment]
@frabbit @nadako make sure to check that in case getCallStack/getExceptionStack is not supported it should return null and not error (a null check is necessary before doing .shift() for instance). Maybe adding unit tests to check at least either null or the number of elements in the stack for callstack/exceptionstack would be nice too :)

**Comment by nadako** [comment]
I added check for null and a very basic unit tests that checks whether `callStack`/`exceptionStack` result is an `Array`. I have no idea how to check more (maybe we could check `.length > 0` on supported platforms though). Is this good enough?

**Comment by nadako** [comment]
waiting for CI to faild hard at this :-)

**Comment by nadako** [comment]
hmm, looks like saucelabs aren't running for PRs, and there are some unrelated errors caused by c++ change. could we merge this and see what happens? :)

**Comment by nadako** [comment]
I rebased this against latest `development` and is now waiting for CI for merging into 3.2. This PR had enough time to mature.

**Comment by nadako** [comment]
Looks good. I also added support for hooking `source-map-support` module into `haxe.CallStack`, so it returns hx positions instead of js ones. Example use:
```haxe
static function __init__() {
    var sm = js.Lib.require("source-map-support");
    sm.install();
    haxe.CallStack.wrapCallSite = sm.wrapCallSite;
}
```
